rpc/signing/middleware.go

```diff
@@ -0,0 +1,82 @@
+package signing
+
+import (
+	"bytes"
+	"io"
+	"net/http"
+
+	"github.com/0xsequence/waas-authenticator/proto"
+)
+
+func Middleware(signer Signer) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if r.Header.Get("Accept-Signature") == "" {
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			var body bytes.Buffer
+			ww := newWrappedRW(w, &body)
+
+			next.ServeHTTP(ww, r)
+
+			b, err := newHTTPSignatureBuilder(signer, &body, r, ww.Header(), ww.status)
+			if err != nil {
+				proto.RespondWithError(w, err)
+				return
+			}
+
+			if err := b.Generate(r.Context()); err != nil {
+				proto.RespondWithError(w, err)
+				return
+			}
+
+			if err := ww.finalize(); err != nil {
+				proto.RespondWithError(w, err)
+			}
+		})
+	}
+}
+
+type wrappedRW struct {
```

Comment:
lets ask Vojtech to review the wrapped writer when he's back, these are always wonky we've experienced over the years

Reply:
@patrislav if the primary goal is to buffer the response body, you could use chi's 
```go
var buf bytes.Buffer
ww := middleware.NewWrapResponseWriter(w, r.ProtoMajor)
ww.Tee(&buf)
```

Reply:
Writing response writers is tricky, 'cause you need to implement all the methods that the parent response writer implements (e.g. `.Flush()`, `.Hijack()`, `Push()`, `ReadFrom()`), so you don't "hide" its capabilities to downstream middleware/handlers.

Reply:
Ok, I thought about this some more and I think the general use-case might be useful for many more things, aka  "Postpone writing to http.ResponseWriter until the full body is known, so we can sign the response with a HTTP header / or so we can post-process the body etc."

1. I propose to extend chi's `middleware.WrapResponseWriter` struct to be able to discard writes to the response writer (while still counting the bytes & writing to `tee`).
    Something like
    ```go
    ww := middleware.NewWrapResponseWriter(w, r.ProtoMajor)
    var buf bytes.Buffer
    ww.Tee(&buf)
    ww.Discard() // discard writes to the original response writer
    
    next.ServeHTTP(wrapWriter, r)
    
    // compute signature from buf.Bytes()
    // set headers
    
    n, err := buf.WriteTo(w) // write to the body to the original response writer
    ```
2. I propose open-sourcing `rpc/signing` middleware as `github.com/go-chi/httpsign` pkg

Reply:
Thank you @VojtechVitek!

1. This looks good and would fit perfectly with our usecase! Do you want me to open a PR with the required changes to `chi`?
2. It would be great, however, this is a very basic implementation and specific to the needs we have in waas-authenticator (e.g. lacks support to sign most headers, customisation is almost non-existent, etc..)

However, with all this process I'm a bit concerned about the timeline - we need this PR merged before the next release, that is next week at the latest, so anything we can do to speed it up would be appreciated! 🙏

Reply:
Sounds good @patrislav

Reply:
With the release of chi 5.1.0 this is now integrated in this PR, @VojtechVitek if you could take one more look?